src/lib/eina/eina_inline_lock_win32.x

```diff
@@ -0,0 +1,767 @@
+   DWORD err = GetLastError();
+   if (ok != 0) return EINA_TRUE;
+   else if ((err == ERROR_NOT_ENOUGH_MEMORY) || (err == ERROR_ACCESS_DENIED)
+            || (err == ERROR_OUTOFMEMORY))
+    {
+       return EINA_FALSE;
+    }
+   else EINA_LOCK_ABORT_DEBUG((int)err, spin_init, spinlock);
+   return EINA_FALSE;
+#else
+   return eina_lock_new(spinlock);
+#endif
+}
+
+static inline void
+_eina_spinlock_free(Eina_Spinlock *spinlock)
+{
+#ifdef EINA_HAVE_WIN32_SPINLOCK
+   DeleteCriticalSection(spinlock);
+   DWORD ok = GetLastError();
+   if (ok != ERROR_SUCCESS) EINA_LOCK_ABORT_DEBUG(ok, spin_destroy, spinlock);
+#else
+   eina_lock_free(spinlock);
+#endif
+}
+
+static inline Eina_Lock_Result
+_eina_spinlock_take_try(Eina_Spinlock *spinlock)
+{
+#ifdef EINA_HAVE_WIN32_SPINLOCK
+   int ok = TryEnterCriticalSection(spinlock);
+   if (ok != 0) return EINA_LOCK_SUCCEED;
+   else if (ok == 0) return EINA_LOCK_FAIL;
+   else
+    {
+       DWORD err = GetLastError();
+       EINA_LOCK_ABORT_DEBUG((int)err, spin_trylock, spinlock);
+    }
+   return EINA_LOCK_FAIL;
+#else
+   return eina_lock_take_try(spinlock);
+#endif
+}
+
+static inline Eina_Lock_Result
+_eina_spinlock_take(Eina_Spinlock *spinlock)
+{
+#ifdef EINA_HAVE_WIN32_SPINLOCK
+   int ok;
+
+#ifdef EINA_HAVE_DEBUG_THREADS
+   if (eina_spinlock_take_try(spinlock) == EINA_LOCK_SUCCEED) return EINA_LOCK_SUCCEED;
+#endif
+
+   for (;;)
+     {
+        ok = EnterCriticalSection(spinlock);
+        if (ok != 0) break;
+        else {
+           DWORD err = GetLastError();
+           EINA_LOCK_ABORT_DEBUG((int)err, spin_lock, spinlock);
+        }
+     }
+
+   return EINA_LOCK_SUCCEED;
+#else
+   return eina_lock_take(spinlock);
+#endif
+}
+
+static inline Eina_Lock_Result
+_eina_spinlock_release(Eina_Spinlock *spinlock)
+{
+#if defined(EINA_HAVE_WIN32_SPINLOCK)
+   LeaveCriticalSection(spinlock);
+   DWORD ok = GetLastError();
+   if (ok == ERROR_SUCCESS) return EINA_LOCK_SUCCEED;
+   else if (ok == ERROR_ACCESS_DENIED) return EINA_LOCK_FAIL;
+   else EINA_LOCK_ABORT_DEBUG((int)ok, spin_unlock, spinlock);
+   return EINA_LOCK_FAIL;
+#else
+   return eina_lock_release(spinlock);
+#endif
+}
+
+static inline Eina_Bool
+_eina_semaphore_new(Eina_Semaphore *sem, int count_init)
+{
+   if (sem && (count_init >= 0))
+     {
+        sem = CreateSemaphoreA(NULL, count_init, count_init, NULL);
+        DWORD ok = GetLastError();
+        if (ok == ERROR_SUCCESS) return EINA_TRUE;
+     }
+   return EINA_FALSE;
+}
+
+static inline Eina_Bool
+_eina_semaphore_free(Eina_Semaphore *sem)
+{
+   if (sem)
+     {
+        CloseHandle(sem);
+        DWORD ok = GetLastError();
+        if (ok == ERROR_SUCCESS) return EINA_TRUE;
+     }
+   return EINA_FALSE;
+}
+
+static inline Eina_Bool
+_eina_semaphore_lock(Eina_Semaphore *sem)
+{
+   if (sem)
+     {
+        DWORD ok;
+        for (;;)
+          {
+             WaitForSingleObject(sem, INFINITE);
+             ok = GetLastError();
```

Comment:
```suggestion
        for (;;)
          {
             WaitForSingleObject(sem, INFINITE);
             DWORD ok = GetLastError();
```

Reply:
But this way I will declare a `DWORD` every iteration of `for`

Reply:
No, you won't. It'll have literally zero impact on generated binary, performance, memory allocated, etc.

Reply:
You can think of a register: if it is allocated in a register, it'll just happen of the same register being reused in every iteration (without any previous initialization).